### PR TITLE
Elasticsearch: Use semver strings to identify ES version

### PR DESCRIPTION
**Pull request by Elfo404**
Instead of using integers to identify an ES version, I switched to using `semver` to allow for greater flexibility when dealing with feature toggles based on configured ES versions.

Given we are adding more features to Elasticsearch, relying on numbers to identify versions and apply logic based on them is becoming cumbersome.
By using `semver` strings we can reduce complexity and improve readability. 

Old versions are automatically coerced to new ones if the data source settings still use the old number format following the existing logic.

**Notes:**
The backend and frontend implementations have different behaviors when dealing with "unknown" version numbers; In the frontend, we are defaulting to `5` (`5.0.0`) while in the backend we rejecting the query (unsupported Elasticsearch version).
We might want to tackle this in a follow-up PR and uniform the behavior.

**Partially unrelated to this change:**
Something I was also considering, is declaring 2 different interfaces for `ElasticsearchOptions` so that one (the current one) will describe historically the shape of these settings (basically whatever we expect the JSON schema to be), and the new one (ie. `NormalizedElasticsearchOptions`) will describe the current shape and types actually used in the data source.
For example, in this case:
```ts
interface ElasticsearchOptions {
  version?: string | number;
}

interface NormalizedElasticsearchOptions {
  version: string;
}
```

IMHO this would be nice to prevent checking for nulls/undefined all over the data source when we know there's a mechanism to normalize those values. For example `ElasticDetails` should only work with a normalized data structure, lifting the responsibility of normalizing that data to (and only) `ConfigEditor`. This way we enforce "legacy checks" to be in one place only and hopefully it will allow us to improve code quality.
Same applies to `ElasticsearchQuery`

### CODE REVIEW

File: package.json [hunk 1]
```diff
@@ -299,6 +300,7 @@
     "rst2html": "github:thoward/rst2html#990cb89",
     "rxjs": "6.6.3",
     "search-query-parser": "1.5.4",
+    "semver": "^7.1.3",
```
Comment (Elfo404) [inline package.json:303]: to avoid any issue I used the same version (and constraint) we have in `@grafana/cli`

**Comment by gabor** [comment]
(the idea seems cool (use semver to simplify version-number-checks), i'm just nitpicking here)

i wonder if some comments should be put somewhere to explain that we use semver for version-checks and nothing more.
because, for example, the `.esVersion` attribute will not contain the real elastic-search version right?

looking at https://github.com/grafana/grafana/blob/d729dc38d956b7ab62333ce7a7691f46860902da/public/app/plugins/datasource/elasticsearch/configuration/ElasticDetails.tsx#L17-L23

it seems what the user chooses is a semver-range-ish-thing, in other words, there is a `^` character before the semver-strings, implied, right?

so this is maybe what i'd address somehow with comments, it's no big deal, i'm just worried that someone later comes to the code, sees `semver`, and will think that things are more semver than they really are (if this makes sense 😄 )

but overall, it seems a reasonable idea.

**Comment by Elfo404** [comment]
@gabor `esVersion` will contain either a version string (`value`) from 
```ts
const esVersions = [ 
   { label: '2.x', value: '2.0.0' }, 
   { label: '5.x', value: '5.0.0' }, 
   { label: '5.6+', value: '5.6.0' }, 
   { label: '6.0+', value: '6.0.0' }, 
   { label: '7.0+', value: '7.0.0' }, 
 ]; 
```
or the old format (`2, 5, 56, 60, 70`).

This doesn't change the behavior in any way compared to the previous version.

there's some documentation update needed to inform users that with provisioning they should start using a version string, what happens with that is they can use whatever version they want (note, a version, not a semver constraint) because we can now actually compare configured versions to known ES version.

so basically:
> it seems what the user chooses is a semver-range-ish-thing, in other words, there is a ^ character before the semver-strings, implied, right?

no, that's not right. They simply choose the version of their ES instance, as explained by the docs in https://grafana.com/docs/grafana/latest/datasources/elasticsearch/#elasticsearch-versions: 

> The value 5.6+ means version 5.6 or higher, but lower than 6.0. The value 6.0+ means version 6.0 or higher, but lower than 7.0. Finally, 7.0+ means version 7.0 or higher, but lower than 8.0.

Does it make sense?

**Comment by gabor** [comment]
> > The value 5.6+ means version 5.6 or higher, but lower than 6.0. The value 6.0+ means version 6.0 or higher, but lower than 7.0. Finally, 7.0+ means version 7.0 or higher, but lower than 8.0.
> 
> Does it make sense?

yes, thanks :+1:

File: pkg/tsdb/elasticsearch/client/client.go [hunk 2]
```diff
@@ -313,7 +340,7 @@ func (c *baseClientImpl) createMultiSearchRequests(searchRequests []*SearchReque
 }
 
 func (c *baseClientImpl) getMultiSearchQueryParameters() string {
-	if c.version >= 70 {
+	if c.version.Major() >= 70 {
```
Comment (ifrost) [inline pkg/tsdb/elasticsearch/client/client.go:343]: Shouldn't it be 7, not 70?

```suggestion
	if c.version.Major() >= 7 {
```

File: pkg/tsdb/elasticsearch/client/search_request_test.go [hunk 3]
```diff
@@ -15,7 +16,8 @@ func TestSearchRequest(t *testing.T) {
 	Convey("Test elasticsearch search request", t, func() {
 		timeField := "@timestamp"
 		Convey("Given new search request builder for es version 5", func() {
-			b := NewSearchRequestBuilder(5, interval.Interval{Value: 15 * time.Second, Text: "15s"})
+			version5, _ := semver.NewVersion("2.0.0")
```
Comment (ifrost) [inline pkg/tsdb/elasticsearch/client/search_request_test.go:19]: The test is for 5 but semver says 2.0.0?

```suggestion
			version5, _ := semver.NewVersion("5.0.0")
```

How come the test is passing? 🤔

File: public/app/plugins/datasource/elasticsearch/datasource.ts [hunk 4]
```diff
@@ -485,7 +486,7 @@ export class ElasticDatasource extends DataSourceApi<ElasticsearchQuery, Elastic
    * search_after feature.
    */
   showContextToggle(): boolean {
-    return this.esVersion > 5;
+    return gte(this.esVersion, '5.0.0');
```
Comment (ifrost) [inline public/app/plugins/datasource/elasticsearch/datasource.ts:489]: It was `>` before, now it's `>=`. Is it intended?

File: public/app/plugins/datasource/elasticsearch/components/QueryEditor/MetricAggregationsEditor/MetricEditor.tsx [hunk 5]
```diff
@@ -39,17 +40,17 @@ const isBasicAggregation = (metric: MetricAggregation) => !metricAggregationConf
...
 ): Array<SelectableValue<MetricAggregationType>> => {
   // we'll include Pipeline Aggregations only if at least one previous metric is a "Basic" one
   const includePipelineAggregations = previousMetrics.some(isBasicAggregation);
 
   return (
     Object.entries(metricAggregationConfig)
       // Only showing metrics type supported by the configured version of ES
-      .filter(([_, { minVersion = 0, maxVersion = esVersion }]) => {
+      .filter(([_, { versionRange = '*' }]) => {
         // TODO: Double check this
```
Comment (ifrost) [inline public/app/plugins/datasource/elasticsearch/components/QueryEditor/MetricAggregationsEditor/MetricEditor.tsx:52]: Does this TODO comment still apply?

**Review by ifrost — CHANGES_REQUESTED** [review]

**Comment by ifrost** [comment]
Just a few inconsistencies. Apart from that it looks elastibloodylicious 😎

File: pkg/tsdb/elasticsearch/client/client.go (hunk 2 above)
Comment (Elfo404) [reply pkg/tsdb/elasticsearch/client/client.go:343]: ABSOLUTELY yes. :D 

thanks for catching it!

File: pkg/tsdb/elasticsearch/client/search_request_test.go (hunk 3 above)
Comment (Elfo404) [reply pkg/tsdb/elasticsearch/client/search_request_test.go:19]: i guess my find-replace skills are not top-notch :)

but I think they are still failing for other reasons

File: public/app/plugins/datasource/elasticsearch/datasource.ts (hunk 4 above)
Comment (Elfo404) [reply public/app/plugins/datasource/elasticsearch/datasource.ts:489]: yeah, it slipped through at some point, `search_after` is actually supported by `5.0.0` : https://www.elastic.co/guide/en/elasticsearch/reference/5.0/search-request-search-after.html

File: public/app/plugins/datasource/elasticsearch/components/QueryEditor/MetricAggregationsEditor/MetricEditor.tsx (hunk 5 above)
Comment (Elfo404) [reply public/app/plugins/datasource/elasticsearch/components/QueryEditor/MetricAggregationsEditor/MetricEditor.tsx:52]: nope, i'll remove it! 👍

File: public/app/plugins/datasource/elasticsearch/datasource.ts (hunk 4 above)
Comment (ifrost) [reply public/app/plugins/datasource/elasticsearch/datasource.ts:489]: nooice

**Comment by Elfo404** [comment]
@gabor regarding

> so this is maybe what i'd address somehow with comments, it's no big deal, i'm just worried that someone later comes to the code, sees semver, and will think that things are more semver than they really are (if this makes sense 😄 )

do you think it would make sense to add a tooltip to the version picker label explaining this in a concise way?

**Comment by gabor** [comment]
> @gabor regarding
> 
> > so this is maybe what i'd address somehow with comments, it's no big deal, i'm just worried that someone later comes to the code, sees semver, and will think that things are more semver than they really are (if this makes sense 😄 )
> 
> do you think it would make sense to add a tooltip to the version picker label explaining this in a concise way?

@Elfo404 
it could help, though i'm not sure how to explain it in a concise way 😿 

or maybe modify the picker-options to work like:
- 5.0.0 - 5.5.x
- 5.6.x
- 6.0.0 - 6.8.x
- 7.0.0 - 7.x.x

though this could get even more confusing maybe :-(

i could imagine a fun gui-widget, where we show a row of radio-buttons like:
`[5.0.0-5.5.x] [5.6.x] [6.0.0-6.8.x] [7.0.0-7.x.x]`
and tell the user to pick the correct one... the idea is to mimic a continuous "range" from where the user picks.

but yeah, ,this is getting more and more wild :-) maybe go with a tooltip?

File: docs/sources/datasources/elasticsearch.md [hunk 6]
```diff
@@ -56,9 +56,12 @@ a time pattern for the index name or a wildcard.
...
-Currently the versions available are `2.x`, `5.x`, `5.6+`, `6.0+` or `7.0+`. The value `5.6+` means version 5.6 or higher, but lower than  6.0. The value `6.0+` means
-version 6.0 or higher, but lower than 7.0. Finally, `7.0+` means version 7.0 or higher, but lower than 8.0.
+Be sure to specify your Elasticsearch version in the version selection dropdown. This is very important as there are differences on how queries are composed and which functionalities are made available in the query editor.
+Currently the versions available are `2.x`, `5.x`, `5.6+`, `6.0+` and `7.0+`. Select the option that best matches your Elasticsearch version.
+
+Grafana assumes you are running the lowest possible version for the specified range so that in case there new features or breaking changes in a future Elasticsearch release you won't be affected.
+
+> **Example** Assuming you are running Elasticsearch `7.6.1` and you selected `7.0+`, if a new feature is made available for Elasticsearch `7.5.0` or newever releases, another option will be added for `7.5+` and you won't be affected until you explicitly set the new option in your settings.
```
Comment (Elfo404) [inline docs/sources/datasources/elasticsearch.md:64]: @hemdrup @gabor clearer?

Comment (achatterjee-grafana) [inline docs/sources/datasources/elasticsearch.md:64]:
```suggestion
You must specify the version of your Elasticsearch data source in the version selection dropdown. Different query compositions and functionalities are available in the query editor for different versions.
Available Elasticsearch versions are `2.x`, `5.x`, `5.6+`, `6.0+`, and `7.0+`. Select the option that best matches your data source version.

Grafana assumes that you are running the lowest possible version for a specified range. This ensures that new features or breaking changes in a future Elasticsearch release will not affect your configuration.

For example, suppose you are running Elasticsearch `7.6.1` and you selected `7.0+`. If a new feature is made available for Elasticsearch `7.5.0` or newer releases, the `7.5+` option will be available.  You will not be affected until you explicitly select the new `7.5+` option in your settings.
```

**Review by achatterjee-grafana — CHANGES_REQUESTED** [review]
Added doc change request.

**Comment by achatterjee-grafana** [comment]
Hi @Elfo404, can you add a milestone to the PR. Thanks!

**Comment by Elfo404** [comment]
@achatterjee-grafana applied the changes, the milestone is already set to 8.0.0

File: docs/sources/datasources/elasticsearch.md [hunk 7]
```diff
@@ -56,9 +56,12 @@ a time pattern for the index name or a wildcard.
 
 ### Elasticsearch version
 
-Be sure to specify your Elasticsearch version in the version selection dropdown. This is very important as there are differences on how queries are composed.
-Currently the versions available are `2.x`, `5.x`, `5.6+`, `6.0+` or `7.0+`. The value `5.6+` means version 5.6 or higher, but lower than  6.0. The value `6.0+` means
-version 6.0 or higher, but lower than 7.0. Finally, `7.0+` means version 7.0 or higher, but lower than 8.0.
+You must specify the version of your Elasticsearch data source in the version selection dropdown. Different query compositions and functionalities are available in the query editor for different versions.
```
Comment (achatterjee-grafana) [inline docs/sources/datasources/elasticsearch.md:59]:
```suggestion
Select the version of your Elasticsearch data source from the version selection dropdown. Different query compositions and functionalities are available in the query editor for different versions.
```

File: docs/sources/datasources/elasticsearch.md [hunk 8]
```diff
@@ -56,9 +56,12 @@ a time pattern for the index name or a wildcard.
...
+Available Elasticsearch versions are `2.x`, `5.x`, `5.6+`, `6.0+`, and `7.0+`. Select the option that best matches your data source version.
+
+Grafana assumes that you are running the lowest possible version for a specified range. This ensures that new features or breaking changes in a future Elasticsearch release will not affect your configuration.
+
+For example, suppose you are running Elasticsearch `7.6.1` and you selected `7.0+`. If a new feature is made available for Elasticsearch `7.5.0` or newer releases, the `7.5+` option will be available.  You will not be affected until you explicitly select the new `7.5+` option in your settings.
```
Comment (achatterjee-grafana) [inline docs/sources/datasources/elasticsearch.md:64]:
```suggestion
For example, suppose you are running Elasticsearch `7.6.1` and you selected `7.0+`. If a new feature is made available for Elasticsearch `7.5.0` or newer releases, then a `7.5+` option will be available.  However, your configuration will not be affected until you explicitly select the new `7.5+` option in your settings.
```

**Review by achatterjee-grafana — CHANGES_REQUESTED** [review]
A few more changes, then we should be good to go. Thanks!

**Review by achatterjee-grafana — APPROVED** [review]
Doc changes look good!

**Comment by marefr** [comment]
Might want to update https://grafana.com/docs/grafana/latest/administration/provisioning/#data-sources in regards to esVersion as well

File: docs/sources/administration/provisioning.md [hunk 9]
```diff
@@ -154,7 +154,7 @@ Since not all datasources have the same configuration settings we only have the
 | maxSeries               | number  | Influxdb                                                         | Max number of series/tables that Grafana processes                                          |
 | httpMethod              | string  | Prometheus                                                       | HTTP Method. 'GET', 'POST', defaults to GET                                                 |
 | customQueryParameters   | string  | Prometheus                                                       | Query parameters to add, as a URL-encoded string.                                                 |
-| esVersion               | number  | Elasticsearch                                                    | Elasticsearch version as a number (2/5/56/60/70)                                            |
+| esVersion               | string  | Elasticsearch                                                    | Elasticsearch version (E.g. `7.0.0`, `7.6.1`)                                            |
```
Comment (gabor) [inline docs/sources/administration/provisioning.md:157]: just to be sure: does this have any backward-compatibility problems?

**Review by gabor — COMMENTED** [review]
code works, i left some nit-picks, they are not critical though.

File: public/app/plugins/datasource/elasticsearch/types.ts [hunk 10]
```diff
@@ -12,7 +12,7 @@ export type Interval = 'Hourly' | 'Daily' | 'Weekly' | 'Monthly' | 'Yearly';
 
 export interface ElasticsearchOptions extends DataSourceJsonData {
   timeField: string;
-  esVersion: number;
+  esVersion: string;
```
[the remark this answers is not in the document]
Comment (Elfo404) [reply public/app/plugins/datasource/elasticsearch/types.ts:15]: it describes the "after" format. in the PR description I also mentioned this problem and I think it would be out of scope for this PR, but I definitely want to iterate over this and have different typings for "whatever might come in" and "what we actually work with internally"

File: pkg/tsdb/elasticsearch/client/search_request_test.go [hunk 11]
```diff
@@ -446,7 +449,8 @@ func TestSearchRequest(t *testing.T) {
 func TestMultiSearchRequest(t *testing.T) {
 	Convey("Test elasticsearch multi search request", t, func() {
 		Convey("Given new multi search request builder", func() {
-			b := NewMultiSearchRequestBuilder(0)
+			version2, _ := semver.NewVersion("2.0.0")
```
[the remark this answers is not in the document]
Comment (Elfo404) [reply pkg/tsdb/elasticsearch/client/search_request_test.go:452]: I'm not sure why it was 0 tbh, I think I blindly replaced it.

looking at the test it feels like there was no reason at all (also because we never supported anything lower than 2: https://github.com/grafana/grafana/pull/33646/files/ebd3f7a33595ce6edff0fec1c0963e0e0ceffb2b#diff-95c1bcdd942efc587e6f08b4175e52e30410f94425bc990a9260fb446f131831L71 ) 

but I'll try to understand better

File: public/app/plugins/datasource/elasticsearch/types.ts (hunk 10 above)
[the remark this answers is not in the document]
Comment (gabor) [reply public/app/plugins/datasource/elasticsearch/types.ts:15]: @Elfo404 just make sure you don't forget that this was once a number :)

**Review by gabor — APPROVED** [review]
code works, semver-checks make it nicer 👍 LGTM!

**Review by ifrost — APPROVED** [review]
